Improved ranking for search results and updated search UI without the artifical delay at the loading screen

kozlovsky opened the pull request:
This PR implements improved search with better item ranking implemented at the DB level, as described in https://github.com/Tribler/tribler/issues/2250#issuecomment-1157735819. Elements with relevant titles and many seeders are placed closer to the top of the search result list.

UI shows local search results immediately, ready without any artificial delay. Remote items show automatically with a slight delay to avoid result flickering and have highlighted background to prevent user confusion when remote results appear on the screen.

Search UI has a progress bar to show the progress for the remote searches. The tooltip of the progress bar displays the number of complete remote requests and the total number of remote requests. It is possible to click on the progress bar to stop processing remote results if they take a long time.

Remote requests have a timeout of 20 seconds to be incorporated into the search result in the UI, and the progress bar actual progress takes into account this hard timeout as well as the current progress on receiving remote results; the more results we have, the close the progress bar to a finish.

drew2a commented:
I guess I've found a QT-related bug:

1. Open the Tribler
2. Search for "ubuntu"
3. Click on a row

<img width="573" alt="image" src="https://user-images.githubusercontent.com/13798583/193239375-5ee75ed5-1f4f-450e-8140-1789309908b0.png">

drew2a commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -1,13 +1,18 @@
 Search utilities.
 
-Author(s): Jelle Roozenburg, Arno Bakker
+Author(s): Jelle Roozenburg, Arno Bakker, Alexander Kozlovsky
```
:)

kozlovsky replied:
We can remove the line with authors altogether. I added the name just because the original people mentioned here do not have a relation to 85% of the code in this file anymore :)

drew2a commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,108 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    total_error = 0
+    for i, term in enumerate(query):
+        # The first word is more important than the second word, and so on
+        term_weight = 5 / (5 + i)
```
It is not clear, why you use value`5`.

kozlovsky replied:
I introduced named constants and added explanation to them

drew2a replied:
Thank you!

drew2a commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,108 @@ def split_into_keywords(string, to_filter_stopwords=False):
+            total_error += skipped * term_weight
+        else:
+            # if the query word is not found in the title, add a big penalty for it
+            total_error += 10 * term_weight
```
It is not clear, why you use value `10` here

kozlovsky replied:
I introduced named constants and added explanation to them

drew2a commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,108 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    total_error += remained_words_error
+
+    # a search rank should be between 1 and 0
+    return 10 / (10 + total_error)
```
`10` should be extracted to a const as it is used multiple times.

kozlovsky replied:
Done

drew2a commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,108 @@ def split_into_keywords(string, to_filter_stopwords=False):
+
+def find_term(term: str, title: Deque[str]) -> Tuple[bool, int]:
+    """
+    Finds the query word in the title.
```
It is unclear what you mean by a "query word". Is it the `term` parameter of the function?

This function looks like function that simply searches the `term` in the `title`. But it contains more advanced logi. It changes the title at the end.

Please add documentation for this or rename/split the function to something more understandable.

kozlovsky replied:
I added a comment with an explanation to the code

drew2a reviewed (changes requested):
Nice improvement! I added a couple of comments for the core logic.

I will check the GUI side immediately after the bugfix is completed.

kozlovsky commented:
> I guess I've found a QT-related bug:

@drew2a thanks, I was able to reproduce it on Mac. On Windows, it does not appear.

synctext commented:
Ambition level is to wrap this up by upcoming Friday.

kozlovsky commented on src/tribler/core/components/metadata_store/remote_query_community/settings.py:
```diff
@@ -5,7 +5,7 @@ class RemoteQueryCommunitySettings(TriblerConfigSection):
     maximum_payload_size: int = 1300
     max_entries: int = maximum_payload_size // minimal_blob_size
-    max_query_peers: int = 5
+    max_query_peers: int = 20
```
I'd like to give a simple answer to this question, but unfortunately, it is a bit complicated.

`max_query_peers` is the number of peers to which `GigaChannelCommunity` (inherited from `RemoteQueryCommunity`) sends remote search requests. Currently, only the `GigaChannelCommunity` subclass of `RemoteQueryCommunity` uses this option ([here](https://github.com/Tribler/tribler/blob/dd8a622f8b8f8be18e0bad24c430b16326e1d0f1/src/tribler/core/components/gigachannel/community/gigachannel_community.py#L223)); the other subclass is `PopularityCommunity` that does not use the option. 

Requests are sent and aggregated entirely in parallel, so as long as on average there are not too many concurrent requests coming to the same peer, there is no reason to limit the number of peers to which we send requests. On average, it is better to send requests to all connected peers to have a variety of results in responses.

Currently, 20 peers is the target number of remote peers for the GigaChannelCommunity. This is because GigaChannelCommunity [uses the RemovePeers strategy](https://github.com/Tribler/tribler/blob/3a49a9ec42396c52e57532ffd5d239e8e94a910c/src/tribler/core/components/gigachannel/gigachannel_component.py#L44), and it [aims to reduce the number of peers to 20](https://github.com/Tribler/tribler/blob/9fe7cf8e3478b9142f2be40c076ce98088021f8e/src/tribler/core/components/gigachannel/community/sync_strategy.py#L16). In PR, I introduce the named constant `TARGET_PEERS_NUMBER` for that number.

At the time of launch, GigaChannelCommunity uses [DiscoverBooster](https://github.com/Tribler/tribler/blob/dd8a622f8b8f8be18e0bad24c430b16326e1d0f1/src/tribler/core/components/gigachannel/community/gigachannel_community.py#L110), thanks to which the number of `GigaChannelCommunity` peers in the first minutes after the launch of Tribler can reach several hundred. I don't think sending parallel search requests to a few hundred peers is necessary a good idea, so I decided to limit the maximum number of concurrent requests, and the number of `TARGET_PEERS_NUMBER` peers looks like a good option.

It is possible to import the `TARGET_PEERS_NUMBER` constant to use it as a value for `max_query_peers` option in `RemoteQueryCommunitySettings`, but it also requires some additional considerations, as we do not import other settings values in a similar way. For now, I added a comment to the `max_query_peers` option that refers to `TARGET_PEERS_NUMBER`.

Now a few words about moving `max_query_peers` to a more appropriate place in settings.

As only `GigaChannelCommunity` uses this option, it is possible to move it from the `RemoteQueryCommunity` settings to the [settings](https://github.com/Tribler/tribler/blob/dd8a622f8b8f8be18e0bad24c430b16326e1d0f1/src/tribler/core/components/gigachannel/community/settings.py#L7) of `GigaChannelCommunity`, but it provokes additional questions and for that reason in my opinion lies outside of the scope of the current PR. Two examples of questions:

1. should we have `RemoteQueryCommunitySettings` at all, as it is essentially an abstract class, and settings should probably be specified in a concrete subclass, to be able to have different settings in different concrete community subclasses (right now, `GigaChannelCommunity` and `PopularityCommunity`)
2. should we put search-related settings in a separate config section, as the search functionality is not essentially tied to `GigaChannelCommunity`?

In my opinion, the option initially was placed in the wrong config section, but moving it to a correct section is a task of another possible PR.

drew2a replied:
Thank you for the explanation 🙏 now `20` is not a magic number anymore.

kozlovsky commented on src/tribler/core/tests/test_search_utils.py:
```diff
@@ -15,3 +21,83 @@ def test_filter_keywords():
+    # Seeders are good for the rank
+    assert torrent_rank(query, 'Big Buck Bunny', seeders=100, freshness=100 * DAY) == pytest.approx(0.876923)
+
+    # The more seeders the better
```
Thanks for the suggestion, I updated the test and it is indeed looks better this way

kozlovsky commented on src/tribler/core/tests/test_search_utils.py:
```diff
@@ -15,3 +21,83 @@ def test_filter_keywords():
+    assert torrent_rank(query, 'Big Buck Bunny II') == pytest.approx(0.80381679)
+
+    # The closer to the start of the string non-matching words are placed in the title, the worse is rank
+    assert torrent_rank(query, 'Big Buck Brown Bunny') == pytest.approx(0.75061099)
```
Good idea!

kozlovsky commented:
@drew2a I fixed the bug you mentioned. When doing this, I changed the way the dynamically added search results items are highlighted.

In UI, I want to highlight dynamically added results with a slightly different color so the user does not become disoriented when new rows suddenly shift the previous content of the search result list.

Initially, I painted just a title column with a slightly different background color for dynamically added results. As your screenshot shows, it looks pretty ugly when the row is selected.

To fix it, I first tried to use Qt [roles](https://doc.qt.io/qt-6/qt.html#ItemDataRole-enum) for model data items. `Qt::BackgroundRole` should allow specifying different background brushes for different table rows, and it sounded like the right way to solve the task.

[`QStyledItemDelegate`](https://doc.qt.io/qt-6/qstyleditemdelegate.html) that we use in Tribler has the [`initStyleOption`](https://doc.qt.io/qt-6/qstyleditemdelegate.html#initStyleOption) method that should allow the background brush to be customizable.

Unfortunately, it turns out that all these Qt mechanics with roles is not working if a Qt widget or its parent has a CSS style. With any CSS style specified, Qt completely ignores all roles from models. As we extensively use CSS styling in Tribler, we cannot use roles to paint different search result items in a different way and cannot specify different background colors for different table rows.

As a workaround, I switched to activating hovered state for a short period for new items. That looks like the only way to specify different backgrounds for some items. And on practice, the resulting highlighting looks quite good to me.

drew2a commented:
@kozlovsky the bug has been fixed. Confirmed. UI looks good to me.

Note, that dynamically added items are not grouped into the snippet:

<img width="567" alt="image" src="https://user-images.githubusercontent.com/13798583/197482562-9fd147aa-d187-4884-a402-caf81266f3c5.png">

kozlovsky commented:
@drew2a suggest to group remote results in a separate PR

drew2a commented on src/tribler/core/components/metadata_store/db/store.py:
```diff
@@ -731,7 +739,11 @@ def get_entries_query(
                     f"""
                     (1 if g.metadata_type == {CHANNEL_TORRENT} else 2 if g.metadata_type == {COLLECTION_NODE} else 3),
-                    desc(g.health.seeders), desc(g.health.leechers)
+                    raw_sql('''search_rank(
+                        $txt_filter, g.title, torrentstate.seeders + 0.1 * torrentstate.leechers,
+                        $int(time()) - strftime('%s', g.torrent_date)
+                    ) DESC'''),
+                    desc(g.health.last_check)  # just to trigger the TorrentState table inclusion into the left join
```
Thank you!

kozlovsky commented on src/tribler/core/tests/test_search_utils.py:
```diff
@@ -15,3 +21,129 @@ def test_filter_keywords():
+
+    # The exact match ranked as pretty high
+
+    r1 = torrent_rank(query, 'Big Buck Bunny')  # 0.81
```
Because 1 is supposed to be the highest rank. For rank to be highest, a torrent should have an exact title match and an infinite number of seeders. So, if a torrent has an ideal title match but zero seeders, it still is not good enough to have the rank 1.

drew2a replied:
Thank you for the explanation.

drew2a commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    find_term("A", deque(["A", "C", "X"])) -> (found=True, skipped=0, rest=deque(["C", "X"]))
+    find_term("B", deque(["C", "X"])) -> (found=False, skipped=0, rest=deque(["C", "X"]))
+    find_term("C", deque(["C", "X"])) -> (found=True, skipped=0, rest=deque(["X"]))
+    Conclusion: huge penalty for missing one query word plus a minor penalty for one excess title word.
```
Thank you for the extensive explanation! Now it is clear what the function does. 🙏 

However, I would suggest you change the design of the function in the following way:
instead of modifying the argument, please return a triple `(found, skipped, rest)`. 
Then the parameters become immutable and the behavior of the function becomes more straightforward.

The current time complexity of the function is `O(n)+O(k)`

One of the options is to rewrite the function to 

```python
def find_term(term: str, title: List[str]) -> Tuple[bool, int, List]:
    accum = []

    for i, word in enumerate(title):
        if word == term:
            rest = title[i + 1::]
            rest.extend(accum)
            return True, i, rest
        else:
            accum.append(word)

    return False, 0, accum
```
It has the same time complexity: `O(n)+O(k)` but a better design.

The tests:

```python
def test_find_term():
    assert find_term("A", ["A", "B", "C"]) == (True, 0, ["B", "C"])
    assert find_term("B", ["B", "C"]) == (True, 0, ["C"])
    assert find_term("C", ["C"]) == (True, 0, [])

    assert find_term("A", ["X", "Y", "A", "B", "C"]) == (True, 2, ["B", "C", "X", "Y"])
    assert find_term("B", ["B", "C", "X", "Y"]) == (True, 0, ["C", "X", "Y"])
    assert find_term("C", ["C", "X", "Y"]) == (True, 0, ["X", "Y"])

    assert find_term("A", ["A", "C", "X"]) == (True, 0, ["C", "X"])
    assert find_term("B", ["C", "X"]) == (False, 0, ["C", "X"])
    assert find_term("C", ["C", "X"]) == (True, 0, ["X"])
```

Ref: https://wiki.python.org/moin/TimeComplexity

kozlovsky replied:
This function is executed directly inside the SQL query and is potentially called for millions of torrents. In this case, the performance of the function directly affects the performance of the SQL query because the SQLite code itself works faster than the Python code executed inside the SQL code.

I did a quick benchmark of the function I use in PR and the function that you are suggesting:

```python
import timeit
t1 = timeit.Timer(
    "calculate_rank(['a', 'b', 'c', 'd', 'e'], ['x', 'a', 'b', 'y', 'z', 'c', 'd', 'e', 'f', 'g'])",
    "from tribler.core.utilities.search_utils import calculate_rank")
t2 = timeit.Timer(
    "calculate_rank2(['a', 'b', 'c', 'd', 'e'], ['x', 'a', 'b', 'y', 'z', 'c', 'd', 'e', 'f', 'g'])",
    "from tribler.core.utilities.search_utils import calculate_rank2")
t1.repeat()
t2.repeat()
```
The results are
```python
[2.4469536999999946, 2.4567987000000358, 2.4902578999999605, 2.5008845999999494, 2.5010604000000285]
```
for the function in the PR and
```python
[3.1153759999999693, 3.1840986999999927, 3.18589350000002, 3.1968425000000025, 3.2020163000000252]
```
for the function that you are suggesting.

So, the function that does not modify the title in place works 28% slower.

It is a pretty big difference and can lead to a 10% performance difference for the entire query. Tribler full-text searches are already slow; making them even slower just to simplify the function slightly is unnecessary. I think that the function in the PR is not so complex, and the user experience is more important here.

drew2a replied:
@kozlovsky could you please give the full source of the test you performed?

kozlovsky replied:
@drew2a sure, you can take it from here: https://pastebin.com/DTK6beT7

It shows the following results to me:
```python
For the test, time one million invocations of function, repeat 5 times
Original function speed: [2.6123462, 2.6423102, 2.6609092, 2.6560733, 2.6591641]
Suggested function speed: [3.365587399999999, 3.3657550999999977, 3.3585778000000026, 3.3598278000000015, 3.353927500000001]
The slowdown is 28.387558279986052 percent
```

drew2a replied:
Thank you. 
Confirmed that your function is faster, so let's keep it.

But then we should rename it to something that reflects the fact that the `title` changes significantly during the function call.

Something `rotate_title_by_word`? I don't like this name, but maybe you will come up with something better.

kozlovsky replied:
I agree that the name should reflect the fact that the function mutates the title.
I renamed it from `find_word` to `find_word_and_rotate_title`

drew2a replied:
🙏

drew2a commented on src/tribler/core/tests/test_search_utils.py:
```diff
@@ -15,3 +21,129 @@ def test_filter_keywords():
+
+    assert r21 < r20 < r19
+
+    # The more seeders is still better
+
+    r22 = torrent_rank(query, 'Buck Bunny', seeders=10, freshness=5 * DAY)  # 0.44805194
+    r23 = torrent_rank(query, 'Buck Bunny', seeders=100, freshness=5 * DAY)  # 0.46821428
+    r24 = torrent_rank(query, 'Buck Bunny', seeders=1000, freshness=5 * DAY)  # 0.4883766
+
+    assert r21 < r22 < r23 < r24
+
+    # The more days from the check the less relevant the number of seeders is
+
+    r25 = torrent_rank(query, 'Buck Bunny', seeders=1000, freshness=10 * DAY)  # 0.48306818
+    r26 = torrent_rank(query, 'Buck Bunny', seeders=1000, freshness=20 * DAY)  # 0.47563636
+
+    assert r26 < r25 < r24
+
+    # The exact match has a good rank
+    r27 = torrent_rank('Sintel', 'Sintel')  # 0.81
+    assert r27 > 0.8
+
+    # Non-matching words at the end of the title give slightly worse results
+    r28 = torrent_rank('Sintel', 'Sintel Part II')  # 0.79553571
+
+    # Non-matching words at the beginning of the title are much worse
+    r29 = torrent_rank('Sintel', 'Part of Sintel')  # 0.664925373
+
+    # Too many non-matching words give a bigger penalty
+    r30 = torrent_rank('Sintel', 'the.script.from.the.movie.Sintel.pdf')  # 0.52105263
+
+    assert r30 < r29 < r28 < r27
+
+    # Some more examples
+
+    r31 = torrent_rank("Internet's Own Boy", "Internet's Own Boy")  # 0.81
+    r32 = torrent_rank("Internet's Own Boy", "Internet's very Own Boy")  # 0.75099337
+    r33 = torrent_rank("Internet's Own Boy", "Internet's very special Boy person")  # 0.4353166986
+
+    assert r33 < r32 < r31
```
I've played with my own search implementation and found that the current test style is not convenient for bug hunting. 

For example, if `torrent_rank` is broken and the assertion `   assert r8 < r7 < r6 < r5 < r1` raises an error, then it is impossible (really hardly possible) to find out which exact`torrent_rank` invocation failed.

You will see in the test the following output:
```python
        # The closer to the start of the string non-matching words are placed in the title, the worse is rank
        r5 = torrent_rank(query, 'Big Buck Bunny II')  # 0.80381679
        r6 = torrent_rank(query, 'Big Buck Brown Bunny')  # 0.75061099
        r7 = torrent_rank(query, 'Big Bad Buck Bunny')  # 0.74242068
        r8 = torrent_rank(query, 'Boring Big Buck Bunny')  # 0.73125
    
>       assert r8 < r7 < r6 < r5 < r1
E       assert 0.72 < 0.670344827586207
``` 
It is really hard to guess which of the four invocations failed. 

To improve the test I suggest restructuring this test into the following:
```suggestion
def test_torrent_rank():
    query = 'Big Buck Bunny'

    # The exact match ranked as pretty high
    full_match = torrent_rank(query, 'Big Buck Bunny')  # 0.81
    assert full_match > 0.8

    # Seeders are good for the rank
    # The more seeders the better
    # The fewer days have passed since the creation of the torrent, the higher its rank
    assert torrent_rank(query, 'Big Buck Bunny', seeders=1000, freshness=1 * DAY) > \
           torrent_rank(query, 'Big Buck Bunny', seeders=1000, freshness=100 * DAY) > \
           torrent_rank(query, 'Big Buck Bunny', seeders=100, freshness=100 * DAY) > \
           full_match

    # If a title contains non-matching words missed in the query string it is not as good as the exact match
    # The closer to the start of the string non-matching words are placed in the title, the worse is rank
    assert full_match > \
           torrent_rank(query, 'Big Buck Bunny II') > \
           torrent_rank(query, 'Big Buck Brown Bunny') > \
           torrent_rank(query, 'Big Bad Buck Bunny') > \
           torrent_rank(query, 'Boring Big Buck Bunny')

    # The more non-matching words are in the title, the worse is rank
    assert full_match > \
           torrent_rank(query, 'Big Buck A Bunny') > \
           torrent_rank(query, 'Big Buck A B Bunny') > \
           torrent_rank(query, 'Big Buck A B C Bunny')

    # Non-matching words close to the beginning of the title give a bigger penalty
    assert full_match > \
           torrent_rank(query, 'Big A Buck Bunny') > \
           torrent_rank(query, 'Big A B Buck Bunny') > \
           torrent_rank(query, 'Big A B C Buck Bunny')

    assert full_match > \
           torrent_rank(query, 'A Big Buck Bunny') > \
           torrent_rank(query, 'A B Big Buck Bunny') > \
           torrent_rank(query, 'A B C Big Buck Bunny')

    assert torrent_rank(query, 'Big A Buck Bunny') > \
           torrent_rank(query, 'A Big Buck Bunny')
    assert torrent_rank(query, 'Big A B Buck Bunny') > \
           torrent_rank(query, 'A B Big Buck Bunny')
    assert torrent_rank(query, 'Big A B C Buck Bunny') > \
           torrent_rank(query, 'A B C Big Buck Bunny')

    # Wrong order of words in the title imposes a penalty to the rank
    assert full_match > \
           torrent_rank(query, 'Big Bunny Buck')

    # Missed query words imposes a really big penalty
    assert torrent_rank(query, 'Big Buck') < 0.5

    # The close the missed words to the beginning of the query, the worse
    assert torrent_rank(query, 'Big Buck') > \
           torrent_rank(query, 'Big Bunny') > \
           torrent_rank(query, 'Buck Bunny')

    # The more seeders is still better
    assert torrent_rank(query, 'Buck Bunny', seeders=1000, freshness=5 * DAY) > \
           torrent_rank(query, 'Buck Bunny', seeders=100, freshness=5 * DAY) > \
           torrent_rank(query, 'Buck Bunny', seeders=10, freshness=5 * DAY) > \
           torrent_rank(query, 'Buck Bunny')

    # The more days from the check the less relevant the number of seeders is
    assert torrent_rank(query, 'Buck Bunny', freshness=5 * DAY) > \
           torrent_rank(query, 'Buck Bunny', freshness=10 * DAY) > \
           torrent_rank(query, 'Buck Bunny', freshness=20 * DAY)

    # The exact match has a good rank
    assert torrent_rank('Sintel', 'Sintel') > 0.8

    # Non-matching words at the end of the title give slightly worse results
    # Non-matching words at the beginning of the title are much worse
    # Too many non-matching words give a bigger penalty
    assert torrent_rank('Sintel', 'Sintel') > \
           torrent_rank('Sintel', 'Sintel Part II') > \
           torrent_rank('Sintel', 'Part of Sintel') > \
           torrent_rank('Sintel', 'the.script.from.the.movie.Sintel.pdf')

    # Some more examples
    assert torrent_rank("Internet's Own Boy", "Internet's Own Boy") > \
           torrent_rank("Internet's Own Boy", "Internet's very Own Boy") > \
           torrent_rank("Internet's Own Boy", "Internet's very special Boy person")
```

After this change, the test output will be the following:

```python
>       assert full_match > \
               torrent_rank(query, 'Big Buck Bunny II') > \
               torrent_rank(query, 'Big Buck Brown Bunny') > \
               torrent_rank(query, 'Big Bad Buck Bunny') > \
               torrent_rank(query, 'Boring Big Buck Bunny')
E       AssertionError: assert 0.670344827586207 > 0.72
E        +  where 0.670344827586207 = torrent_rank('Big Buck Bunny', 'Big Buck Brown Bunny')
E        +  and   0.72 = torrent_rank('Big Buck Bunny', 'Big Bad Buck Bunny')
```

kozlovsky replied:
Thanks, I rewrited the tests according to your suggestion

drew2a replied:
🙏

drew2a commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+RANK_NORMALIZATION_COEFF = 10
+
+
+def calculate_rank(query: List[str], title: List[str]) -> float:
+    """
+    Calculate the similarity of the title to the query as a float value in range [0, 1].
+    """
+    if not query:
+        return 1.0
+
+    if not title:
+        return 0.0
+
+    title = deque(title)
+    total_error = 0
+    for i, term in enumerate(query):
+        # The first word is more important than the second word, and so on
+        term_weight = POSITION_COEFF / (POSITION_COEFF + i)
+
+        # Read the description of the `find_term` function to understand what is going on. Basically, we are trying
+        # to find each query word in the title words, calculate the penalty if the query word is not found or if there
+        # are some title words before it, and then rotate the skipped title words to the end of the title. This way,
+        # the least penalty got a title that has query words in the proper order at the beginning of the title.
+        found, skipped = find_term(term, title)
+        if found:
+            # if the query word is found in the title, add penalty for skipped words in title before it
+            total_error += skipped * term_weight
+        else:
+            # if the query word is not found in the title, add a big penalty for it
+            total_error += MISSED_WORD_PENALTY * term_weight
+
+    # a small penalty for excess words in the title that was not mentioned in the search phrase
+    remainder_weight = 1 / (REMAINDER_COEFF + len(query))
+    remained_words_error = len(title) * remainder_weight
+    total_error += remained_words_error
+
+    # a search rank should be between 1 and 0
+    return RANK_NORMALIZATION_COEFF / (RANK_NORMALIZATION_COEFF + total_error)
```
I am wondering... could `calculate_rank` be replaced by something simpler?

After a quick research, I prepared a test function that satisfies @synctext requirements posted [here](https://github.com/Tribler/tribler/issues/2250#issuecomment-1157735819) 

```python
def test_calculate_rank():
    # rule 2
    assert torrent_rank(query='Sintel', title='Sintel') > \
           torrent_rank(query='Sintel', title='the.script.from.the.movie.Sintel.pdf')

    # rule 3
    assert torrent_rank(query='Sintel', title='Sintel-Part-I') > \
           torrent_rank(query='Sintel', title='Part-of-Sintel')

    # rule 4
    assert torrent_rank(query="The Internet's Own Boy", title="The Internet's Own Boy") > \
           torrent_rank(query="The Internet's Own Boy", title="The Internet's Own Boy: The Story") >\
           torrent_rank(query="The Internet's Own Boy", title="The Internet's Own Boy: The Story of Aaron Swartz")

    # rule 5
    assert torrent_rank(query="Internet's Own Boy", title="Internet's Own Boy - Aaron Swartz") > \
           torrent_rank(query="Internet's Own Boy", title="Internet's very Own Boy")

    assert torrent_rank(query="Internet's Own Boy", title="Internet's Own Boy - Aaron Swartz") > \
           torrent_rank(query="Internet's Own Boy", title="Internet's very special Boy person")
```

And came up with the simple `calculate_rank` function that is probably good enough for the first iteration of search improvement.

The new version:
```python
def calculate_rank(query: List[str], title: List[str]) -> float:
    title = ' '.join(title)
    query = ' '.join(query)
    matcher = SequenceMatcher(None, query, title, autojunk=False)
    longest = matcher.find_longest_match(0, len(query), 0, len(title))
    return matcher.quick_ratio() * longest.size
```

The old version:
```python
def calculate_rank(query: List[str], title: List[str]) -> float:
    if not query:
        return 1.0

    if not title:
        return 0.0

    title = deque(title)
    total_error = 0
    for i, term in enumerate(query):
        # The first word is more important than the second word, and so on
        term_weight = POSITION_COEFF / (POSITION_COEFF + i)

        # Read the description of the `find_term` function to understand what is going on. Basically, we are trying
        # to find each query word in the title words, calculate the penalty if the query word is not found or if there
        # are some title words before it, and then rotate the skipped title words to the end of the title. This way,
        # the least penalty got a title that has query words in the proper order at the beginning of the title.
        found, skipped = find_term(term, title)
        if found:
            # if the query word is found in the title, add penalty for skipped words in title before it
            total_error += skipped * term_weight
        else:
            # if the query word is not found in the title, add a big penalty for it
            total_error += MISSED_WORD_PENALTY * term_weight

    # a small penalty for excess words in the title that was not mentioned in the search phrase
    remainder_weight = 1 / (REMAINDER_COEFF + len(query))
    remained_words_error = len(title) * remainder_weight
    total_error += remained_words_error

    # a search rank should be between 1 and 0
    return RANK_NORMALIZATION_COEFF / (RANK_NORMALIZATION_COEFF + total_error)


def find_term(term: str, title: Deque[str]) -> Tuple[bool, int]:
    try:
        skipped = title.index(term)
    except ValueError:
        return False, 0

    title.rotate(-skipped)  # rotate skipped words to the end
    title.popleft()  # remove found word
    return True, skipped
```

drew2a replied:
If I got you correctly, you have the following concerns about the suggested function:
1. The output is not normalized to [0..1]
2. It works worse for one corner case: "the excess words at the beginning and the end of the title".

About the (1) -- yes, it is a draft. The output could be easily normalized. I didn't do it because for the prototype it is not necessary.

About (2) yes, but the suggested solution works pretty well for the most cases and extremely easy in terms of implementation.

Pros and cons in the single table:
|   | Original Solution | Suggested Solution |
| ------------- | ------------- |------------- |
| Johan Test  | ✅ | ✅ |
| Speed  | 1.4 | 20.8 |
| LOC  | 100 | 6 |
| Kozlovsky Test (corner cases)  | ✅ | ❌ |

The speed test has been performed by the similar test that you did before:

```python
import timeit
from difflib import SequenceMatcher
from typing import List


def calculate_rank2(query: List[str], title: List[str]) -> float:
    title = ' '.join(title)
    query = ' '.join(query)
    matcher = SequenceMatcher(None, query, title, autojunk=False)
    longest = matcher.find_longest_match(0, len(query), 0, len(title))
    return matcher.quick_ratio() * longest.size


t1 = timeit.Timer(
    "calculate_rank('The Big Buck Bunny', 'Making of The Big Buck Bunny')",
    "from tribler.core.utilities.search_utils import calculate_rank\n")

t2 = timeit.Timer(
    "calculate_rank2('The Big Buck Bunny', 'Making of The Big Buck Bunny')",
    "from __main__ import calculate_rank2\n")

print('For the test, time 10k invocation of function, repeat 5 times')

function1_results = t1.repeat(repeat=5, number=10000)
print(f'Original function speed: {function1_results}')

function2_results = t2.repeat(repeat=5, number=10000)
print(f'Suggested function speed: {function2_results}')

# Use the fastest result from each benchmark as it is the least affected by other parallel tasks and so the most precise
slowdown = min(function2_results) * 100 / min(function1_results) - 100
print(f'The slowdown is {slowdown} percent')
```

To summarize the above my opinion is that we should keep your solution despite its complexity and size 🤝

kozlovsky replied:
Actually, the performance test requires a small fix. On my machine, the test shows that the function based on `SequenceMatcher` is 1579% slower than the solution from the PR. But the `calculate_rank` function actually expects lists of words, not strings, in its arguments. So, the correct timers definitions should look like this:

```python
t1 = timeit.Timer(
    "calculate_rank('The Big Buck Bunny'.split(), 'Making of The Big Buck Bunny'.split())",
    "from tribler.core.utilities.search_utils import calculate_rank\n")

t2 = timeit.Timer(
    "calculate_rank2('The Big Buck Bunny'.split(), 'Making of The Big Buck Bunny.split()')",
    "from __main__ import calculate_rank2\n")
```

With this change, the function based on SequenceMatcher is still slower, but only 849% slower.

kozlovsky replied:
> the suggested solution works pretty well for the most cases

Actually, I consider the remaining cases quite significant and not so rare. It is pretty possible that Tribler has an almost perfect title match in the database, but it receives a heavy penalty for additional words at the end of the title name and, as a result, is not included in the search result list. With the function implementation from the PR, such results will not be accidentally penalized, and users will see them. I expect it to be important for around 10% of search queries.

kozlovsky replied:
In other words, if a user searches for "The Big Buck Bunny 1080p", and the title in the database is "The Big Buck Bunny MP4 1080p 30FPS Eng", it should almost not be penalized for a non-perfect match, and the algorithm from the PR takes that into account.

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    """
+    Calculates rank based on the number of seeders. The result is normalized to the range [0, 1]
+    """
+    return seeders / (100 + seeders)  # inf seeders -> 1; 100 seeders -> 0.5; 10 seeders -> approx 0.1
```
The plot above with the seeders range [-400, 400] is misleading. The number of seeders is a positive value, usually in the range [0, 1000].

I want the `seeders_rank` to return a value in the range [0, 1]. It should return 0 for a torrent with zero seeders and 1 for a torrent with an infinite number of seeders.

So, the actual graph for a function looks like this. With zero seeders, the torrent has `seeders_rank` 0; with 100 seeders it goes up to 0.5, and then slowly increases to 1.0 when the number of seeders reaches infinity. It does not have singularities like on your previous graph.

<img width="1148" alt="Screenshot 2022-10-24 at 17 33 08" src="https://user-images.githubusercontent.com/2931136/197566230-9bcdf225-aff7-466c-b2ad-80af6b4d42d5.png">

It is possible to use different curves with the similar shape: `x / (N + x)`, `2 * arctan(x / N) / PI`, `1 - exp(x * ln(0.5) / N)`, but it is not actually matter on practice

<img width="1189" alt="Screenshot 2022-10-24 at 18 34 56" src="https://user-images.githubusercontent.com/2931136/197578739-aebb2a83-b4af-4623-afbc-244962e3243b.png">

I use the simplest and most efficient function with desired properties: zero rating for zero seeders, 0.5 rating for 100 seeders, 1 rating for an infinite number of seeders, and soft curve.

drew2a replied:
Thank you for the explanation. 
It would be nice if they were placed in the code. It follows from your explanation that `100` is not a randomly picked number, but a number with meaning. 
That's why it also may be reflected in the code:

```python
def seeders_rank(seeders: float, half_point: float = 100) -> float:
    """ Calculates rank based on the number of seeders.

    This function has desired properties: zero rating for zero seeders, 0.5 rating for `half_point` seeders,
    1 rating for an infinite number of seeders, and a soft curve.
    
    The result is normalized to the range [0, 1].

    Args:
        seeders:  The number of seeders. It is a positive value, usually in the range [0, 1000]
        half_point: The number of seeders at which the rank should return half of the maximum value, i.e. 0.5

    Returns: float in the range [0, 1]
    """

    # It is possible to use different curves with the similar shape: 
    # 2 * arctan(x / N) / PI, 
    # 1 - exp(x * ln(0.5) / N)
    # but it does not actually matter in practice
    return seeders / (half_point + seeders)
```

kozlovsky replied:
I added the description to the formula and extracted a `SEEDERS_HALF_RANK` named constant

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    return torrent_rank(query, title, seeders + leechers * 0.1, freshness)
+
+
+def torrent_rank(query: str, title: str, seeders: int = 0, freshness: Optional[float] = 0) -> float:
```
I added comments about the freshness parameter and the ouput rank range

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    return result
+
+
+def seeders_rank(seeders: float) -> float:
```
I changed it to `int` for clarity and add `leechers` as a second parameter

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    """
+    Calculates rank based on the number of seeders. The result is normalized to the range [0, 1]
+    """
+    return seeders / (100 + seeders)  # inf seeders -> 1; 100 seeders -> 0.5; 10 seeders -> approx 0.1
```
Done

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+
+def freshness_rank(freshness: Optional[float] = 0):
+    """
+    Calculates rank based on the torrent freshness. The result is normalized to the range [0, 1]
```
I added a description to the code. Freshness is the number of seconds since the torrent creation. It is more convenient to use than a timestamp, as, with freshness, it is not necessary to calculate the current time inside the function, and tests are a bit easier.

I have seen a few torrents in the database whose timestamps are in the future, and their calculated freshness is negative. If a freshness value is less or equal to zero, it is considered to be invalid, and the resulting rank is zero.

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+
+    days = (freshness or 0) / SECONDS_IN_DAY
+
+    return 1 / (1 + days / 30)  # 2x drop per 30 days
```
I added a description to a function it is conceptually similar to the previous functions I described above.     # The function declines from 1.0 to 0.0 on range (0..Infinity], with the following properties:
 * for just created torrents the rank value is close to 1.0
 * for 30-days old torrents the rank value is 0.5
 * for very old torrens the rank value is going to zero
 
 
![image](https://user-images.githubusercontent.com/2931136/199231032-3b03495d-975c-4933-a3e4-65895152709b.png)

 
The exact shape of the curve is not important, we could use exponential curve as well, but the current formula calculates a bit faster

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
     return [kw for kw in keywords if len(kw) > 0 and kw not in DIALOG_STOPWORDS]
+
+
+def item_rank(query: str, item: dict):
```
I added comments to functions and corresponding tests: test_title_rank_range, test_freshness_rank_range, test_seeders_rank_range, test_torrent_rank_range

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+REMAINDER_COEFF = 10
+
+# The exact value of this coefficient is not important. It is used to convert total_error value to a rank value.
+# The total_error value is some positive number. We want to have the resulted rank in range [0, 1].
```
I added the tests: test_title_rank_range, test_freshness_rank_range, test_seeders_rank_range, test_torrent_rank_range

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    The `term` parameter is a word from a search string. It is called `term` and not `word` because it can also be
+    a stemmed version of the word if the comparison algorithm implemented in the top-level `torrent_rank` function
+    works with stemmed words. The ability to work with stemmed words was added to `torrent_rank` and then removed,
+    as it currently does not give significant benefits, but it can be added again in the future.
```
I changed `term` to `word`

drew2a replied:
🙏

kozlovsky commented on src/tribler/core/utilities/search_utils.py:
```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+    Example 6, query "A B C", title "A C X":
+    find_term("A", deque(["A", "C", "X"])) -> (found=True, skipped=0, rest=deque(["C", "X"]))
+    find_term("B", deque(["C", "X"])) -> (found=False, skipped=0, rest=deque(["C", "X"]))
+    find_term("C", deque(["C", "X"])) -> (found=True, skipped=0, rest=deque(["X"]))
```
Done

drew2a replied:
🙏
